File: Startup.R
# Input Variables (x) = R.D Spend , Administration , Marketing Spend , State
# Output Variable(y) = Profit

# Importing the dataset
Startup <- read.csv(file.choose())
colnames(Startup) <- c("RD","Admin","MS","State","Profit") # RD = R.D.Spend , Admin = Administrartion & MS = Marketing Spend
View(Startup)

# Creating dummy variables for State

install.packages("dummies")
library(dummies)

str(Startup)
Startup$State <- as.factor(Startup$State)
Startup$State = as.numeric(Startup$State)
str(Startup)
View(Startup)
attach(Startup)

# Normal distribution
qqnorm(RD)
qqline(RD)

# Exploratory data analysis:
# 1. Measures of central tendency
# 2. Measures of dispersion
# 3. Third moment business decision
# 4. Fourth moment business decision
# 5. Probability distributions of variables 
# 6. Graphical representations (Histogram, Box plot, Dot plot, Stem & Leaf plot, Bar plot, etc.)

summary(Startup)

# Graphical representations (Histogram, Box plot, Dot plot, Stem & Leaf plot, Bar plot, etc.)

# Boxplot Representation

boxplot(Startup$RD, col = "dodgerblue4",main = "R.D.Spend")
boxplot(Startup$Admin, col = "dodgerblue4",main = "Administration")
boxplot(Startup$MS, col = "dodgerblue4",main = "Marketing Spend")
boxplot(Startup$Profit, col = "red", horizontal = T,main = "Profit")

# Histogram Representation

hist(Startup$RD,col = "orange", main = "R.D.Spend" )
hist(Startup$Admin,col = "orange", main = "Administration")
hist(Startup$MS,col = "orange", main = "Marketing Spend")
hist(Startup$Profit,col = "red", main = "Profit")

# Scatter plot
plot(RD,Profit) # Plot relation ships between each X with Y
plot(Admin, Profit)
plot(MS,Profit)
#plot(State,Profit) #this shows error as it is a categorical variable

# Or make a combined plot
#pairs(Start_up)   #  doesnt work as there is a categorical variable
#Scatter plot for all pairs of variables
plot(Startup)

cor(RD, Profit) #0.973
cor(Startup) # correlation matrix
#Seems Strong coorelation between profit and R.D.Spend

# The Linear Model of interest
model.Startup <- lm(Profit ~ RD + Admin + MS + State, data = Startup) # lm(Y ~ X)
summary(model.Startup)
# p value shows insignificant for Admin,MS,State
#Over all R square is 0.9507 and p is less than 0.05

model.Admin <- lm(Profit ~ Admin)
summary(model.Admin) # p value is 0.1622 which is greater than 0.05 
#and r square value is also less 0.04029 which makes this insignificant

model.MS <- lm(Profit ~ MS) # pverall p values are significant so we can consider MS
summary(model.MS)

model.State <- lm(Profit ~ State)
summary(model.State) #P values making state insignificant

model.AMS <- lm(Profit ~ Admin + MS + State )
summary(model.AMS) #state the significant
#Coefficients:
#Estimate Std. Error t value Pr(>|t|)    
#(Intercept)    1.903e+04  1.843e+04   1.033   0.3072    
#Admin          3.239e-01  1.335e-01   2.426   0.0193 *  
# MS             2.507e-01  3.135e-02   7.997 3.48e-10 ***
#StateFlorida  -1.704e+03  9.338e+03  -0.182   0.8561 (insignificant)
#StateNew York  3.876e+03  9.003e+03   0.431   0.6689  (insignificant)

#### Scatter plot matrix with Correlations inserted in graph
# install.packages("GGally")
library(GGally)
ggpairs(Startup)


### Partial Correlation matrix
install.packages("corpcor")
library(corpcor)
cor(Startup) #no strong correlatoin between the inputs

cor2pcor(cor(Startup)) # RD and Profit has strong correaltion

# Diagnostic Plots
install.packages("car")
library(car)

plot(model.Startup) # Residual Plots, QQ-Plot, Std. Residuals vs Fitted, Cook's distance

qqPlot(model.startup, id.n = 5 ,main = "Influential Observations") # QQ plots of studentized residuals, helps identify outliers

# Deletion Diagnostics for identifying influential obseravations
influenceIndexPlot(model.startup, id.n = 3) # Index Plots of the influence measures
influencePlot(model.Startup, id.n = 3 , main = "Influential Observations") # A user friendly representation of the above
#50  and 49 inputs are influential observations

# Regression after deleting the 50th observation
model.Startup1 <- lm(Profit ~ RD + Admin + MS + State, data = Startup[-50, ])
summary(model.Startup1)
# still admin and state seems to be insignificant
model.Startup1 <- lm(Profit ~ RD + Admin + MS + State, data = Startup[-49, ])
summary(model.Startup1)
View(Startup)

#still no use

### Variance Inflation Factors
vif(model.Startup)  # VIF is > 10 => collinearity

#There is no VIF greater than 10  so there is no collinearity problem in this model

# Regression model to check R^2 on Independent variales
VIFRD <- lm(RD ~ Admin + MS + State) # state insignificant
VIFAdmin <- lm(Admin ~ RD + MS + State)# state insignificant
VIFMS <- lm(MS ~ RD + Admin + State)#state insignificant
#VIFState <- lm(State ~ RD + Admin + MS)

summary(VIFRD)
summary(VIFAdmin)
summary(VIFMS)
#summary(VIFState)

# VIF of RDS
1/(1-0.59) # <10 no multi collinearity double confirmed

#### Added Variable Plots ######
avPlots(model.Startup, id.n = 2, id.cex = 0.8, col = "red")

#As per the entire above analysis , state variable results in insignificant

# Linear Model without State
model.final <- lm(Profit ~ RD + Admin + MS, data = Startup)
summary(model.final)

# Linear model without State and influential observation
model.final1 <- lm(Profit ~ RD + Admin + MS, data = Startup[-50, ])
summary(model.final1)


# Added Variable Plots
avPlots(model.final1, id.n = 2, id.cex = 0.8, col = "red")
# admin is insignificant

# Variance Influence Plot
vif(model.final1)

# Evaluation Model Assumptions
plot(model.final1)
plot(model.final1$fitted.values, model.final1$residuals)

qqnorm(model.final1$residuals)
qqline(model.final1$residuals)

# Data Partitioning
n <- nrow(Startup)
n1 <- n * 0.7
n2 <- n - n1
train <- sample(1:n, n1)
test <- Startup[-train, ]

# Model Training
model <- lm(Profit ~ RD + Admin + MS, Startup[train, ])
summary(model)
# r square is 0.946 and overall p values is good but Admin is still insignificant

model <- lm(Profit ~ RD + MS, Startup[train, ])
summary(model)

pred <- predict(model, newdata = test)
actual <- test$Profit
error <- actual - pred

test.rmse <- sqrt(mean(error**2))
test.rmse

train.rmse <- sqrt(mean(model$residuals**2))
train.rmse

# Step AIC
install.packages("MASS")
library(MASS)
stepAIC(model.Startup)


# Still it is overfit need to do regularization.

File: Startup.py
# Input Variables (x) = R&D Spend, Admin, Marketing Spend, State.
# Output Variable(y) = Profit

import pandas as pd
import scipy 
from scipy import stats
import statsmodels.formula.api as sm
from statsmodels.formula.api import ols
import numpy as np

# Import Dataset

Startup = pd.read_csv("C:/Users/SHAJIUDDIN MOHAMMED/Desktop/Data sets/50_Startups.csv")
Startup.columns = "RD","Admin","MS","State","Profit"

#Exploratory Data Analysis

Startup.describe()

import matplotlib.pyplot as plt

# Graphical Representation of Box plot and Histogram

#RD

plt.bar(height = Startup.RD , x = np.arange(1,51,1))
plt.boxplot(Startup.RD)
plt.hist(Startup.RD)

#MS

plt.bar(height = Startup.MS, x = np.arange(1,51,1))
plt.boxplot(Startup.MS)
plt.hist(Startup.MS)


# Joint Plots

import seaborn as sns
sns.jointplot(x = Startup.RD ,y = Startup.Profit)
sns.jointplot(x = Startup.Admin, y = Startup.Profit)
sns.jointplot(x = Startup.MS, y = Startup.Profit)

# Creation of Dummy Variables

print(Startup.info())
Startup.boxplot('Profit','State')

# Checking of N/A values

cat_Startup = Startup.select_dtypes(include = ['object']).copy()
cat_Startup.head()
print(cat_Startup.isnull().values.sum()) 

# if there are N/A values then we need do imputation(remove)
print(cat_Startup['State'].value_counts())

# Plot for State Variable

State_count = cat_Startup['State'].value_counts()
sns.set(style = "darkgrid")
sns.barplot(State_count.index,State_count.values,alpha = 0.9)
plt.title('State influence on Profit')
plt.ylabel('count of States', fontsize=12)
plt.xlabel('State', fontsize=12)
plt.show()

# One hot Encoding Technique to create dummies for categorical varaibles

cat_Startup_onehot_sklearn = cat_Startup.copy()
cat_Startup_onehot_sklearn

from sklearn.preprocessing import LabelBinarizer

lb = LabelBinarizer()
lb_results = lb.fit_transform(cat_Startup_onehot_sklearn['State'])
lb_results_df = pd.DataFrame(lb_results, columns=lb.classes_)

print(lb_results_df.head())

# concate the dummy variable to the data sheet

Startup_df = pd.concat([Startup, lb_results_df], axis=1)
Startup_df
Startup_df = Startup_df.drop(['State'], axis=1)
Startup_df

# Scatter Plot

sns.pairplot(Startup_df.iloc[:, :])
                             
# Correlation matrix 
Startup_df.corr()

# MODEL BUILDING

model = sm.ols('Profit ~ RD + Admin + MS + State', data = Startup).fit()
model.summary()

# From the output the overall R^2 = 0.951 and State and Admin is still insignifcant

# Checking of Influential Values by ploting

import statsmodels.api as smf
smf.graphics.influence_plot(model)

# As the graph shows that 49 index has high influence ,so we can exclude that entire row and creating a new model
Startup_new = Startup.drop(Startup.index[[49]])

# New Model Building without Outliers

model1 = sm.ols('Profit ~ RD + Admin + MS + State', data = Startup_new).fit()
model1.summary()

# By elimanating the outliers the R^2 value has been increased to 0.962 from 0.951 
# And we still see variables show insignificant

# Check for Colinearity to decide to remove a variable using VIF
# Assumption: VIF > 10 = colinearity
# calculating VIF's values of independent variables

rsq_RD = sm.ols('RD ~ MS + Admin + State', data = Startup).fit().rsquared  
vif_RD = 1/(1 - rsq_RD) 
vif_RD


rsq_MS = sm.ols('MS ~ RD + State + Admin', data = Startup).fit().rsquared  
vif_MS = 1/(1 - rsq_MS)
vif_MS 


rsq_Adm = sm.ols('Admin ~ State + RD + MS', data = Startup).fit().rsquared  
vif_Adm = 1/(1 - rsq_Adm) 
vif_Adm # 1.177

# All the above variable VIF values are < 10,So there is no influence

# calculating vif with dummy variables

rsq_state1 = sm.ols('California ~ Admin + RD + MS', data = Startup_df).fit().rsquared  
vif_state1= 1/(1 - rsq_state1) 
vif_state1 # 1.03

rsq_state2 = sm.ols('Florida ~ Admin + RD + MS', data = Startup_df).fit().rsquared  
vif_state2= 1/(1 - rsq_state2) 
vif_state2 # 1.05

rsq_state3 = sm.ols('NewYork ~ Admin + RD + MS', data = Startup_df).fit().rsquared  
vif_state3= 1/(1 - rsq_state3) 
vif_state3 # 1.03


# The above vif analysis shows there is no collinearity between any of the inputs and making state insignificant

# Final Model

model.final = sm.ols('Profit ~ RD + Admin + MS', data = Startup).fit()
model.final.summary()

# Result : still Admin is insignificant so removing admin and checking

model.final1 = sm.ols('Profit ~ RD + MS', data = Startup).fit()
model.final1.summary()

# Now the MS significance is also improved and r square is 0.950

# Prediction
pred = model.final1.predict(Startup)

# Q-Q plot
res = model.final1.resid
smf.qqplot(res)
plt.show()

# Q-Q plot
from scipy import stats
import pylab
stats.probplot(res, dist = "norm", plot = pylab)
plt.show()

# Residuals vs Fitted plot
sns.residplot(x = pred, y = Startup.Profit, lowess = True)
plt.xlabel('Fitted')
plt.ylabel('Residual')
plt.title('Fitted vs Residual')
plt.show()

smf.graphics.influence_plot(model.final1)


### Splitting the data into train and test data 
from sklearn.model_selection import train_test_split
Startup_train, Startup_test = train_test_split(Startup, test_size = 0.2) # 20% test data

# preparing the model on train data 
model_train = sm.ols("Profit ~ RD + MS ", data = Startup_train).fit()

# prediction on test data set 
test_pred = model_train.predict(Startup_test)

# test residual values 
test_resid = test_pred - Startup_test.Profit
# RMSE value for test data 
test_rmse = np.sqrt(np.mean(test_resid * test_resid))
test_rmse # 6615.18

# train_data prediction
train_pred = model_train.predict(Startup_train)

# train residual values 
train_resid  = train_pred - Startup_train.Profit
# RMSE value for train data 
train_rmse = np.sqrt(np.mean(train_resid * train_resid))
train_rmse # 9377.49

# Over fit need to go with regularization

